### PR TITLE
GIVCAMP-269 GIVCAMP-206: Loading, Cache, and 404.

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Trying out new cache controls
- Added loading page
- New settings don't seem to be showing the loading page after all

# Steps to test

1. View https://deploy-preview-180--giving-campaign.netlify.app/ in Safari
2. View loading page https://deploy-preview-180--giving-campaign.netlify.app/loading to preview what the loading will look like
3. Navigate back to the home page by clicking the logo
4. Navigate to a new page using an in page link
5. Click the back button
6. Wait for page to load and scroll around
7. See no back button issue
8. Navigate to  https://deploy-preview-180--giving-campaign.netlify.app/this-page-does-not-exist
9. See 404 page
10. Log in to Netlify and view the [function logs](https://app.netlify.com/sites/giving-campaign/logs/functions/?scope=deploypreview:180)
11. Open up the SSR and ISR function logs
12. Set the time to 'last hour'
13. Wait and you shouldn't see any logs related to pages

For bonus points, make an edit on Storyblok and then 'rebuild' the Netlify environment then check to see if the content shows up.

### CODE REVIEW

File: app/layout.tsx
Comment (sherakama) [inline app/layout.tsx:12]: Static render EVERYTHING

File: app/(storyblok)/[[...slug]]/layout.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/layout.tsx:7]: I figured I may as well add this here too.

**Comment by yvonnetangsu** [comment]
@sherakama It looks like the back button issue on Safari is back 😂 However, I think it worked ok before this commit was pushed up:
https://github.com/SU-SWS/ood-giving-campaign/pull/180/commits/d92d534df80aa0d3ebff7a1c6aa00e1d66e69a84

File: app/(storyblok)/[[...slug]]/page.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/page.tsx:89]: Screenshot 2023-12-20 at 13 35 36
Screenshot 2023-12-20 at 13 48 04

4mb of page data????

Comment (yvonnetangsu) [reply app/(storyblok)/[[...slug]]/page.tsx:89]: What?? 👀

Comment (yvonnetangsu) [reply app/(storyblok)/[[...slug]]/page.tsx:89]: Did a lighthouse run the other day and performance was like 86% 👀

Comment (sherakama) [reply app/(storyblok)/[[...slug]]/page.tsx:89]: Yikes, yeah.

All of the relations (Nested blocks, stories, media, etc) are all fully resolved and well, they add up.

**Comment by yvonnetangsu** [comment]
@sherakama I found something interesting today. It looks like the updates that are published on Storyblok now show up almost instantly - I've tested a few times today. LMK if you're seeing the same or am I going crazy 😂  Not on this PR's preview build - try it on [dev](https://dev--giving-campaign.netlify.app/)

**Comment by sherakama** [comment]
> @sherakama I found something interesting today. It looks like the updates that are published on Storyblok now show up almost instantly - I've tested a few times today. LMK if you're seeing the same or am I going crazy 😂 Not on this PR's preview build - try it on [dev](https://dev--giving-campaign.netlify.app/)

Oh, that probably means ISR is on and is fetching 'live' for each URL...

**Comment by yvonnetangsu** [comment]
> > @sherakama I found something interesting today. It looks like the updates that are published on Storyblok now show up almost instantly - I've tested a few times today. LMK if you're seeing the same or am I going crazy 😂 Not on this PR's preview build - try it on [dev](https://dev--giving-campaign.netlify.app/)
> 
> Oh, that probably means ISR is on and is fetching 'live' for each URL...

Yeah definitely look like it's fetching live

**Comment by sherakama** [comment]
Yep. My guess is that `revalidate: 0` is what's up here.

**Comment by sherakama** [comment]
Screenshot 2023-12-20 at 14 00 44

**Comment by sherakama** [comment]
@yvonnetangsu Added steps to test. This is really ready to test now... not that you haven't already been testing, but you know for realz this time.

File: app/loading/page.tsx
Comment (sherakama) [inline app/loading/page.tsx:1]: Delete this file before merging. This is just a preview.

**Comment by yvonnetangsu** [comment]
@sherakama Looks like this preview is pulling old Storyblok data 😂 For example, on the homepage, this should say "How Daniel Ho...." after an update that was made yesterday
Screenshot 2023-12-20 at 2 49 28 PM

But in this preview, it still says "How RegLab..."
Screenshot 2023-12-20 at 2 50 33 PM

Loading page looks good 👍🏼

**Comment by sherakama** [comment]
I'll pick back up on this now that we are back.

**Comment by yvonnetangsu** [comment]
> I'll pick back up on this now that we are back.

Thanks, @sherakama ! 😄 Happy New Year again 🎉

**Comment by yvonnetangsu** [comment]
404 again 😂

**Comment by sherakama** [comment]
https://answers.netlify.com/t/401-unauthorized-trying-to-download-404-page/95492/11

**Comment by yvonnetangsu** [comment]
> https://answers.netlify.com/t/401-unauthorized-trying-to-download-404-page/95492/11

Oooo

File: app/(storyblok)/[[...slug]]/layout.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/layout.tsx:7]: Wasn't helping anything so this can be removed.

File: app/(storyblok)/[[...slug]]/page.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/page.tsx:23]: This appeared to be helping the Safari back button bug but I moved a few other things so I have to revalidate (see what I did there) that this change is doing the trick.

File: app/(storyblok)/[[...slug]]/page.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/page.tsx:52]: The Nextjs runtime plugin from Netlify keeps a cache but I don't know what's in it, or know if I can see what is in it. This, however, seems to do the job and invalidates all the pages on build time. 

Steps to test this are:
1. Review the home page in the browser
2. Make an edit and publish a change to the home page in storyblok
3. Rebuild the netlify environment through a webhook or the 'retry' option in the netlify UI. If you are going the retry deploy in netlify UI option, DO NOT select the clear cache and deploy option. Just retry.

File: app/(storyblok)/[[...slug]]/page.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/page.tsx:61]: GetAll is supposed to get all the pages but it seemed to be missing the pages from `test/test-stories` without this.

File: app/(storyblok)/editor/page.tsx
Comment (sherakama) [inline app/(storyblok)/editor/page.tsx:29]: We're not pre-rendering anything on this page. Good bye.

File: app/(storyblok)/editor/page.tsx
Comment (sherakama) [inline app/(storyblok)/editor/page.tsx:30]: This overrides the dynamic setting as far as I can tell but both are set here.

File: app/(storyblok)/not-found.tsx
Comment (sherakama) [inline app/(storyblok)/not-found.tsx:1]: Not needed.

File: app/404/page.tsx
Comment (sherakama) [inline app/404/page.tsx:1]: The Netlify plugin seems to be handling this

File: app/global-error.tsx
Comment (sherakama) [inline app/global-error.tsx:1]: These were causing problems and not sure why.

File: app/not-found.tsx
Comment (sherakama) [inline app/not-found.tsx:10]: Not needed.

File: components/StoryPoC/ChatbotStory.tsx
Comment (sherakama) [inline components/StoryPoC/ChatbotStory.tsx:112]: Extra log

File: components/TypeForm/PopOver.tsx
Comment (sherakama) [inline components/TypeForm/PopOver.tsx:4]: Typeform package update moved the export.

File: components/TypeForm/PopUp.tsx
Comment (sherakama) [inline components/TypeForm/PopUp.tsx:3]: Typeform package update moved the export.

File: netlify.toml
Comment (sherakama) [inline netlify.toml:31]: I temporarily removed this as the 404 breaks when it is behind basic auth on Netlify. See former comment url.

File: utilities/getPageMetadata.ts
Comment (sherakama) [inline utilities/getPageMetadata.ts:1]: This wasn't being used.

File: utilities/getPageMetadata.ts
Comment (sherakama) [inline utilities/getPageMetadata.ts:83]: Export was complaining about missing this.

**Comment by sherakama** [comment]
@yvonnetangsu Thanks for following along at home. Through all the hilarity I think this is up and ready for review again.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu Thanks for following along at home. Through all the hilarity I think this is up and ready for review again.

Thank you for the thorough investigation, @sherakama - really appreciate your expertise! I'll review after I finish up that SAA decanter ticket 😄

File: netlify.toml
Comment (sherakama) [reply netlify.toml:31]: https://answers.netlify.com/t/401-unauthorized-trying-to-download-404-page/95492/11

**Comment by yvonnetangsu** [comment]
Hi @sherakama 😄  So everything on your steps to test check out - not seeing the "Loading" anymore which is great  ✅ . No back button issues - tested on both desktop and mobile ✅ . No function log ✅  

However, the "bonus" bit at the end still seems a bit weird. So last night I updated the alt text on some of the images in one story - 

If you look at the same story in dev (which is pulling live data), and inspect this image, you'll see that the latest alt is "An illustration with Yi Cui and elements from his work"
https://dev--giving-campaign.netlify.app/stories/battery-super-power
Screenshot 2024-01-05 at 2 02 05 PM

This morning, I merged dev into this branch which triggered a rebuild of this preview. After that, I inspect this preview, and found that the alt didn't seem to update to the latest published one (like the one above)
https://deploy-preview-180--giving-campaign.netlify.app/stories/battery-super-power
Screenshot 2024-01-05 at 2 01 49 PM

**Comment by yvonnetangsu** [comment]
Please feel free to merge this PR in first, and open another for the cache issue 😄

**Comment by sherakama** [comment]
ok, fun new discovery. 

The Safari scroll bug is back for me. But only when the dev tools are closed.

**Comment by yvonnetangsu** [comment]
> ok, fun new discovery.
> 
> The Safari scroll bug is back for me. But only when the dev tools are closed.

That is strangely specific 😬

**Comment by sherakama** [comment]
Ok, the prefetch removal seems to be helping the safari bug. 

The home page is where I think the issue is as it 500s by default. Gotta think of a way to patch this up.
Without the prefetch, you see the loading page and it is a bit jarring IMHO.

**Comment by sherakama** [comment]
Netlify doesn't (seem) to support on-demand revalidation. I thought I read somewhere today that it won't work but I can't find the link right now.

https://github.com/netlify/next-runtime#incremental-static-regeneration-isr

**Comment by sherakama** [comment]
https://answers.netlify.com/t/netlify-support-for-next-js-caching/102986/8

> _Sometime in January_

On demand revalidation soon I guess. For now, we are stuck with a timeout/rebuild. 

Playing around with this some more.

File: package.json
Comment (sherakama) [inline package.json:23]: [Nextjs DataCache](https://nextjs.org/docs/app/building-your-application/caching#data-cache) is persistent between builds through the `.next/cache` and Netlify plugin. That is why between builds we were not seeing updates to the content. This little hack removes the cached files persistent between deployments.

File: next.config.js
Comment (sherakama) [inline next.config.js:26]: In order to avoid the 500 error on the index page RSC fetch `/.rsc?pvg24` I needed to put the homepage somewhere it could be prefetched (at its slug). This redirect will allow the home link of `/home` to be prefetched but redirect to the correct URL.

File: netlify.toml
Comment (sherakama) [inline netlify.toml:4]: Added a cache clear before build in this new script.

File: components/Logo/LogoLockup.tsx
Comment (sherakama) [inline components/Logo/LogoLockup.tsx:48]: Home link is now to the slug. A redirect makes sure the end user doesn't see it. This is to support the RSC prefetching and Safari where the back button broke when RSC prefetching returned a 500 error. 

The Safari bug is also tied to the ODB Function/ISR functionality somehow but I haven't troubleshooted it and am at a 'it works don't touch it' place.

File: app/(storyblok)/[[...slug]]/page.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/page.tsx:23]: Bug in Safari when revalidate is removed or set to false.

**Comment by sherakama** [comment]
@yvonnetangsu,

This is, once again, ready for your review.

File: next.config.js
Comment (sherakama) [inline next.config.js:5]: This is meant to be a fix for Kube containers but I figured it would help here too.
https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#self-hosting-isr

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu,
> 
> This is, once again, ready for your review.

I hope you know I really appreciate that you're doing all this exploration @sherakama ! 😄

**Comment by sherakama** [comment]
> I hope you know I really appreciate that you're doing all this exploration @sherakama ! 😄

You're most welcome. I enjoy really digging into systems and figuring them out. This one is taking longer than it should have for me but hey, that's what it is.

**Comment by yvonnetangsu** [comment]
@sherakama The publishing content all worked as expected! (I rebuild without clearing cache on Netlify and it does show the latest updates in Storyblok), but I got this Netlify crash when I went to the newsletter sign up page 😂 
Screenshot 2024-01-08 at 4 58 17 PM
Screenshot 2024-01-08 at 4 58 48 PM

**Comment by sherakama** [comment]
> @sherakama The publishing content all worked as expected! (I rebuild without clearing cache on Netlify and it does show the latest updates in Storyblok), but I got this Netlify crash when I went to the newsletter sign up page 😂 Screenshot 2024-01-08 at 4 58 17 PM Screenshot 2024-01-08 at 4 58 48 PM

Can you reproduce this issue? Does it just happen one time after the site is built?

**Comment by yvonnetangsu** [comment]
> > @sherakama The publishing content all worked as expected! (I rebuild without clearing cache on Netlify and it does show the latest updates in Storyblok), but I got this Netlify crash when I went to the newsletter sign up page 😂 !
> Can you reproduce this issue? Does it just happen one time after the site is built?

I can't seem to reproduce it. It happened when I refreshed the sign up page. I've been doing more clicky and reload and haven't seen that again. Shall we merge this? I tested on mobile and everything looks good 👍🏼 😄

**Comment by sherakama** [comment]
> Shall we merge this?

Your call. If you feel this is working and want to run it in dev for a bit before go-live then push the green button but you can also let this bake a bit if you want to hammer on it some more.

**Comment by sherakama** [comment]
Oh, might want to add the deploy-preview context back into the basic auth first. Be aware that 404 pages will break while the auth is enabled.

File: netlify.toml
Comment (sherakama) [inline netlify.toml:31]:
```suggestion
    context = ['branch-deploy', 'deploy-preview', 'production']
```

**Review by yvonnetangsu — APPROVED** [review]
Let's merge it why not 😂  If we see the error again we can come back and re-evaluate. Merging this would allow us to more easily discover any issues too. Thank you @sherakama !!

**Comment by sherakama** [comment]
Woot!

**Comment by yvonnetangsu** [comment]
Nice GIF @sherakama 😆  I've been playing with the site on my phone last night and haven't seen another Netlify function error yet 🤞🏼

**Comment by yvonnetangsu** [comment]
@sherakama It happened one more time just now, but that was _right after_ I published a change on Storyblok. I went to the dev-- site to looked at the story I just edited. After that reloading no issue at all. We can keep an eye on it 👀 
Screenshot 2024-01-09 at 11 08 47 AM

**Pull request merged**
